pandas/core/algorithms.py

```diff
@@ -134,7 +134,7 @@ def _ensure_data(values, dtype=None):
         return values, dtype, 'int64'
 
     # we have failed, return object
-    values = np.asarray(values)
+    values = np.asarray(values, dtype=np.object)
```

Comment:
so we actually should prob use ``pandas.core.dtypes.cast.construct_1d_array_preserving_na`` which is even better here. further pls run the performance suite for things like factorize, value_counts, isin, this a very performance sensitive section.

Reply:
@jreback Actually, [pandas.core.dtypes.cast.construct_1d_ndarray_preserving_na](https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/cast.py#L1245) would not work for two reasons:

 1. For  `[42, 's']`  it returns `array(['42', 's'], dtype='<U11')` rather the wanted `array([42, 's'], dtype=object))`, not sure this is the intended behavior of the function though
 2. For `[np.nan]` it returns `array([nan], dtype=float64)` which leads to `result[0] is np.nan` being False, but we would like to keep the id of the object.